modules/system/keyboard/default.nix

```diff
@@ -0,0 +1,300 @@
+          });
+        }
+      else if (cfg.enableKeyMapping && cfg.mappings != null) then (listToAttrs (map
+        ({ mappings
+         , productId
+         , vendorId
+         , ...
+         }: (nameValuePair "keyboard-${toString productId}" ({
+          serviceConfig.ProgramArguments = [
+            # Use xpc_set_event_stream_handler to mark this event as "consumed",
+            # otherwise it the script will never stop being called (something
+            # like every 10 seconds).
+            "${xpc_set_event_stream_handler}/bin/xpc_set_event_stream_handler"
+            "${pkgs.writeScriptBin "apply-keybindings" (
+                let intToHexString = value:
+                  pkgs.runCommand "${toString value}-to-hex-string"
+                      { } ''printf "%#0x" ${toString value} > $out''; in
+                ''
+                  #!${pkgs.stdenv.shell}
+                  set -euxo pipefail
+
+                  # Sometimes it takes a moment for the keyboard to be
+                  # visible to hidutil, even when the script is launched
+                  # with "LaunchEvents".
+                  function retry () {
+                    local attempt=1
+                    local max_attempts=10
+                    local delay=0.2
+
+                    while true; do
+                      "$@" && break || {
+                        if (test $attempt -lt $max_attempts); then
+                          attempt=$((attempt + 1))
+                          sleep $delay
+                        else
+                          exit 1
+                        fi
+                      }
+                    done
+                  }
+
+                  function get_vendor_id () {
+                    hidutil list --matching keyboard |
+                      awk '{ print $1 }' |
+                      grep $(<${intToHexString vendorId})
+                  }
+
+                  function get_product_id () {
+                    hidutil list --matching keyboard |
+                      awk '{ print $2 }' |
+                      grep $(<${intToHexString productId})
+                  }
+
+                  echo "$(date) configuring keyboard ${toString productId} ($(<${intToHexString productId}))..." >&2
+
+                  retry get_vendor_id
+                  retry get_product_id
+
+                  hidutil property --matching '${builtins.toJSON { ProductID = productId; }}' --set '${mkUserKeyMapping mappings}' > /dev/null
+                ''
+                )}/bin/apply-keybindings"
+          ];
+          serviceConfig.LaunchEvents = {
+            "com.apple.iokit.matching" = {
+              "com.apple.usb.device" = {
+                IOMatchLaunchStream = true;
+                IOProviderClass = "IOUSBDevice";
+                idProduct = productId;
+                idVendor = vendorId;
+              };
+            };
+          };
+          serviceConfig.RunAtLoad = true;
+        })
+        ))
```

Comment:
IMO this is strange formatting but it's what nixpkgs-fmt does and I don't care enough to stop using it